### PR TITLE
 [FLINK-27737] Remove legacy support for unfenced executor in FencedRpcEndpoint

### PR DESCRIPTION
**Pull request by Aitozi**
## What is the purpose of the change

This PR is mean to clean up all the un-fenced executor support in the rpcEndpoint. 

## Brief change log

  - Make the FencedRpcEndpoint's `fencingToken` `@NonNull` and `final`
  - Remove the `Unfencedmessage` and `UnfencedMainThreadExecutor`
  - Remove the `PermanentlyFencedRpcEndpoint` since the FencedRpcEndpoint's token is unchanged now
  - Remove the unnecessary tests.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

### CODE REVIEW

**Comment by Aitozi** [comment]
cc @xintongsong

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/FencedRpcEndpoint.java [hunk 1]
```diff
@@ -40,23 +35,17 @@
  */
 public abstract class FencedRpcEndpoint<F extends Serializable> extends RpcEndpoint {
 
-    private final UnfencedMainThreadExecutor unfencedMainThreadExecutor;
-    private volatile F fencingToken;
+    private final F fencingToken;
     private volatile MainThreadExecutor fencedMainThreadExecutor;
```
Comment (xintongsong) [inline flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/FencedRpcEndpoint.java:39]: It seems we no longer need this `fencedMainThreadExecutor`, as now it is no different from `RpcEndpoint#mainThreadExecutor`.

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/FencedRpcEndpoint.java [hunk 2]
```diff
@@ -65,29 +54,14 @@ protected FencedRpcEndpoint(
                         mainThreadExecutable, this::validateRunsInMainThread, endpointId));
     }
 
-    protected FencedRpcEndpoint(RpcService rpcService, @Nullable F fencingToken) {
+    protected FencedRpcEndpoint(RpcService rpcService, @Nonnull F fencingToken) {
```
Comment (xintongsong) [inline flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/FencedRpcEndpoint.java:57]: We don't usually use `@Nonnull`, as anything not `@Nullable` is expected non-null.

**Review by xintongsong — CHANGES_REQUESTED** [review]
Thanks @Aitozi, the change already look quite good. My major comment is to further remove `FencedRpcEndpoint#fencedMainThreadExecutor`, which should allow us to get rid of pretty much everything but the `fencingToken` in `FencedRpcEndpoint`.

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/FencedRpcEndpoint.java (hunk 2 above)
Comment (Aitozi) [reply flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/FencedRpcEndpoint.java:57]: Get, will remove it

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/FencedRpcEndpoint.java (hunk 1 above)
Comment (Aitozi) [reply flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/FencedRpcEndpoint.java:39]: Nice suggestion, will remove it now.

**Review by xintongsong — APPROVED** [review]
Thanks for addressing the comment. LGTM.

**Pull request closed without merging**